src/persist/src/indexed/columnar.rs

```diff
@@ -0,0 +1,259 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! A columnar representation of ((Key, Val), Time, Diff) data suitable for in-memory
+//! reads and persistent storage.
+
+use std::iter::FromIterator;
+
+use serde::{Deserialize, Serialize};
+
+/// A set of ((Key, Val), Time, Diff) records stored in a columnar representation.
+///
+/// Note that the data are unsorted, and unconsolidated (so there may be multiple
+/// instances of the same ((Key, Val), Time), and some Diffs might be zero, or
+/// add up to zero).
+#[derive(Clone, Debug, Default, Eq, PartialEq, Serialize, Deserialize)]
+pub struct ColumnarRecords {
```

Comment:
thoughts on manually implementing Debug so it's actually useful? i lean toward yes. something like making this one in terms of Ref and implementing Ref with `fmt.debug_list`. we can leave Iter derived since I think iterators don't usually print all their contents as part of their debug impls

Reply:
I implemented them both in terms of `debug_list` because I couldn't find a way to tell the formatter "print an instance of this type which also implements debug", other than the struct builder thingy, which didn't seem right. the complexity of doing that is a tiny bit higher than reusing the columnarrecordref's debug impl, but both debug impls seem simple enough. at any rate, this seems like something we can touch up in post

Reply:
something like `fmt::Debug::fmt(self.borrow(), f)` should work as the ColumnarRecords impl

Reply:
:facepalm: ah yeah ofc. ty fixed!